### PR TITLE
Integration enable/disable refactor

### PR DESCRIPTION
**Pull request by bsneed**
- Modified timeline to determine enablement by a combination of event data as well as settings data.
- Added method `hasIntegrationSettings`.
- Removed confusing state entries re integrations vs. settings for integrations.
- Modified Segment destination to construct integration object as it's being sent to segment rather than try to do it beforehand.

### CODE REVIEW

File: Sources/Segment/Settings.swift [hunk 1]
```diff
@@ -68,12 +68,9 @@ public struct Settings: Codable {
         return integrationSettings(forKey: plugin.key)
     }
     
-    public func isDestinationEnabled(key: String) -> Bool {
+    public func hasIntegrationSettings(forPlugin plugin: DestinationPlugin) -> Bool {
```
Comment (migs647) [inline Sources/Segment/Settings.swift:71]: I like this rename

File: Sources/Segment/Analytics.swift [hunk 2]
```diff
@@ -111,6 +111,15 @@ extension Analytics {
         return settings
     }
     
+    /// Manually enable a destination plugin.  This is useful when a given DestinationPlugin doesn't have any Segment tie-ins at all.
+    /// This will allow the destination to be processed in the same way within this library.
+    /// - Parameters:
+    ///   - plugin: The destination plugin to enable.
+    public func manuallyEnableDestination(plugin: DestinationPlugin) {
```
Comment (migs647) [inline Sources/Segment/Analytics.swift:147]: I'm going back and forth if we want this to be a "core" public API. It makes sense for this to be separated by an extension to me. Thoughts?

File: Sources/Segment/Timeline.swift [hunk 3]
```diff
@@ -210,18 +210,28 @@ extension DestinationPlugin {
...
+        if let disabled: Bool = integrations?.value(forKeyPath: KeyPath(self.key)), disabled == false {
+            customerDisabled = true
+        }
+        
+        var hasSettings = false
+        if let settings = analytics?.settings() {
+            hasSettings = settings.hasIntegrationSettings(forPlugin: self)
+        }
+        
+        return (hasSettings && customerDisabled == false)
```
Comment (migs647) [inline Sources/Segment/Timeline.swift]: I'd prefer this line as `hasSettings == true && customerDisabled == false` OR `hasSettings && !customerDisabled` so we aren't mixing semantics.

**Comment by migs647** [comment]
Liking this direction

**Review by oscb — APPROVED** [review]
Looks good to me!

**Review by alanjcharles — APPROVED** [review]
my brain hurts but it looks good and make sense

File: Sources/Segment/Plugins/SegmentDestination.swift [hunk 4]
```diff
@@ -156,6 +140,29 @@ public class SegmentDestination: DestinationPlugin {
...
 
+// MARK: - Utility methods
+extension SegmentDestination {
+    internal func configureCloudDestinations<T: RawEvent>(event: T) -> T {
+        guard let integrationSettings = analytics?.settings() else { return event }
+        guard let plugins = analytics?.timeline.plugins[.destination]?.plugins as? [DestinationPlugin] else { return event }
+        guard let customerValues = event.integrations?.dictionaryValue else { return event }
+        
+        // take the customer values first.
+        var merged = customerValues
```
Comment (prayansh) [inline Sources/Segment/Plugins/SegmentDestination.swift]: i think we should use integrationSettings as the base value, and then add the customer values to override here

File: Sources/Segment/Timeline.swift [hunk 5]
```diff
@@ -210,18 +210,28 @@ extension DestinationPlugin {
         }
         return result
     }
+    
+    internal func isDestinationEnabled(integrations: JSON?) -> Bool {
```
Comment (prayansh) [inline Sources/Segment/Timeline.swift]: maybe we should change the input parameter to `RawEvent` so that the api isnt misused with any kind of JSON

**Review by prayansh — APPROVED** [review]
looks good. Added some comments for more clarification

File: Sources/Segment/Types.swift [hunk 6]
```diff
@@ -277,14 +277,13 @@ extension RawEvent {
...
         
-        guard let system: System = store.currentState() else { return self }
         guard let userInfo: UserInfo = store.currentState() else { return self }
         
         result.anonymousId = userInfo.anonymousId
         result.userId = userInfo.userId
         result.messageId = UUID().uuidString
         result.timestamp = Date().iso8601()
-        result.integrations = system.integrations
+        result.integrations = try? JSON([String: Any]())
```
Comment (oscb) [inline Sources/Segment/Types.swift:286]: Don't you need to leave the `self.integrations` here? Else it strips all user defined integrations before the timeline processes it

Comment (bsneed) [reply Sources/Segment/Types.swift:286]: Yeah; this happens right when the event comes in.  The user hasn't had a chance to do anything yet so there's nothing to override.

**Pull request merged**